deployment/ccip/changeset/messaging_test.go

```diff
@@ -0,0 +1,209 @@
+			[]byte("hello FeeQuoter"),
+		)
+	})
+
+	t.Run("message to contract implementing CCIPReceiver", func(t *testing.T) {
+		out = runMessagingTestCase(
+			messagingTestCase{
+				testCaseSetup: setup,
+				replayed:      out.replayed,
+				nonce:         out.nonce,
+			},
+			state.Chains[destChain].Receiver.Address(),
+			[]byte("hello CCIPReceiver"),
+			func(t *testing.T) {
+				iter, err := state.Chains[destChain].Receiver.FilterMessageReceived(nil)
+				require.NoError(t, err)
+				require.True(t, iter.Next())
+				// MessageReceived doesn't emit the data unfortunately, so can't check that.
+			},
+		)
+	})
+
+	t.Run("message to contract implementing CCIPReceiver with low exec gas", func(t *testing.T) {
+		out = runMessagingTestCase(
+			messagingTestCase{
+				testCaseSetup: setup,
+				replayed:      out.replayed,
+				nonce:         out.nonce,
+			},
+			state.Chains[destChain].Receiver.Address(),
+			[]byte("hello CCIPReceiver with low exec gas"),
+			func(t *testing.T) {
+				// Message should not be emitted, not enough gas to emit log.
+				// TODO: this is still returning a log, probably the older one since FAILURE is the execution state.
+				// Not enough ctx in the message received log to confirm that it's from another test.
+				// Maybe check the log block number and assert that its < the header before block number from above?
+				// iter, err := ccipReceiver.FilterMessageReceived(&bind.FilterOpts{
+				// 	Start: headerBefore.Number.Uint64(),
+				// })
+				// require.NoError(t, err)
+				// require.False(t, iter.Next(), "MessageReceived should not be emitted in this test case since gas is too low")
+			},
+		)
+	})
+}
+
+func sleepAndReplay(t *testing.T, e ccipdeployment.DeployedEnv, sourceChain, destChain uint64) {
+	time.Sleep(30 * time.Second)
+	replayBlocks := make(map[uint64]uint64)
+	replayBlocks[sourceChain] = 1
+	replayBlocks[destChain] = 1
+	ccipdeployment.ReplayLogs(t, e.Env.Offchain, replayBlocks)
+}
+
+func runMessagingTestCase(
+	tc messagingTestCase,
+	receiver common.Address,
+	msgData []byte,
+	extraAssertions ...func(t *testing.T),
+) (out messagingTestCaseOutput) {
+	// check latest nonce
+	latestNonce, err := tc.onchainState.Chains[tc.destChain].NonceManager.GetInboundNonce(&bind.CallOpts{
+		Context: tests.Context(tc.t),
+	}, tc.sourceChain, tc.sender)
+	require.NoError(tc.t, err)
+	require.Equal(tc.t, tc.nonce, latestNonce)
+
+	startBlocks := make(map[uint64]*uint64)
+	seqNum := ccipdeployment.TestSendRequest(tc.t, tc.deployedEnv.Env, tc.onchainState, tc.sourceChain, tc.destChain, false, router.ClientEVM2AnyMessage{
+		Receiver:     common.LeftPadBytes(receiver.Bytes(), 32),
+		Data:         msgData,
+		TokenAmounts: nil,
+		FeeToken:     common.HexToAddress("0x0"),
+		ExtraArgs:    nil,
+	})
+	expectedSeqNum := make(map[uint64]uint64)
+	expectedSeqNum[tc.destChain] = seqNum
+
+	// hack
+	if !tc.replayed {
+		sleepAndReplay(tc.t, tc.deployedEnv, tc.sourceChain, tc.destChain)
```

Comment:
I'm curious about what's the behavior of the test without the replay? Messages never get committed?

Reply:
Yeah the replay is required, I believe its because we send the messages before the plugins are fully running (including chain readers)